Implement Seek::stream_position() for BufReader

t-rapp opened the pull request:
Optimization over `BufReader::seek()` for getting the current position without flushing the internal buffer.

Related to #31100. Based on the code in #70577.

rust-highfive commented:
r? @kennytm

(rust_highfive has picked a reviewer for you, use r? to override)

LukasKalbertodt commented on src/libstd/io/buffered.rs:
```diff
@@ -257,6 +257,41 @@ impl<R: Seek> BufReader<R> {
+    #[unstable(feature = "seek_convenience", issue = "59359")]
+    pub fn stream_position(&mut self) -> io::Result<u64> {
+        self.inner.stream_position().map(|pos| pos - (self.cap - self.pos) as u64)
+    }
```
I think you meant to put that into the `impl<R: Seek> Seek for BufReader<R>` block instead?

t-rapp replied:
Oh, yes indeed. Shall I add a new commit to fix this or squash the change into the existing one?

LukasKalbertodt replied:
I'd prefer squashing/amending in this case, but however you like!

bors commented:
:umbrella: The latest upstream changes (presumably #73265) made this pull request unmergeable. Please resolve the merge conflicts.

t-rapp commented on src/libstd/io/buffered.rs:
```diff
@@ -383,6 +383,40 @@ impl<R: Seek> Seek for BufReader<R> {
+    /// afterwards will yield the underlying reader at the same position. Use
+    /// [`BufReader::seek`] instead if you require that guarantee.
+    ///
+    /// [`BufReader::seek`]: struct.BufReader.html#method.seek
```
Ok, will remove the link definition.

t-rapp commented:
Updated the implementation to check for subtraction underflow / overflow. Realized that besides a buggy inner reader, this situation can also occur when misusing the (unfortunately public) `BufReader::get_mut()` method. Added a test for that case.

kennytm commented:
r? @LukasKalbertodt

jyn514 commented on library/std/src/io/buffered.rs:
```diff
@@ -383,6 +383,51 @@ impl<R: Seek> Seek for BufReader<R> {
+    ///
+    /// This function will panic if the position of the inner reader is smaller
+    /// than the amount of buffered data. That can happen if the inner reader
+    /// has a buggy implementation of [`Seek::stream_position`], or if position
```
```suggestion
    /// has an incorrect implementation of [`Seek::stream_position`], or if the position
```

jyn514 commented on library/std/src/io/buffered.rs:
```diff
@@ -1208,6 +1253,49 @@ mod tests {
+        let inner = reader.get_mut();
+        assert!(inner.seek(SeekFrom::Start(0)).is_ok());
+        // overflow when subtracting the remaining buffer size from current position
+        let _ = reader.stream_position();
```
This test doesn't do what it intends. It will panic if _any_ of the assertions fail, not just the last one. Instead of using `should_panic`, you could use `catch_unwind` on the last assert.

t-rapp replied:
I see. Replacing the last line to
```
let result = panic::catch_unwind(panic::AssertUnwindSafe(|| reader.stream_position().ok()));
assert!(result.is_err());
```
seems to allow the `should_panic` attribute to be removed and the test to succeed.

LukasKalbertodt commented on library/std/src/io/buffered.rs:
```diff
@@ -1208,6 +1254,49 @@ mod tests {
+
+        // cause internal buffer to be filled but read only partially
+        let mut buffer = [0, 0];
+        assert!(reader.read(&mut buffer).is_ok());
```
```suggestion
        assert!(reader.read_exact(&mut buffer).is_ok());
```

While very unlikely, `read` is allowed to only read one byte into `buffer`.

t-rapp replied:
Ok, fixed in latest push.

LukasKalbertodt reviewed (changes requested):
Sorry for the delay. But looks almost ready to merge now! Just a tiny thing I noticed.

bors commented:
:umbrella: The latest upstream changes (presumably #72808) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #75979) made this pull request unmergeable. Please resolve the merge conflicts.

LukasKalbertodt commented on library/std/src/io/buffered/tests.rs:
```diff
@@ -86,6 +87,49 @@ fn test_buffered_reader_seek_relative() {
 }
 
+#[test]
+#[unstable(feature = "seek_convenience", issue = "59359")]
```
The unstable attribute does not need to be applied to tests. This line can be removed.

t-rapp replied:
I think I've seen this somewhere back when writing the tests but don't remember anymore where it was.
Will remove the unstable test attributes.

LukasKalbertodt commented on library/std/src/io/buffered/tests.rs:
```diff
@@ -86,6 +87,49 @@ fn test_buffered_reader_seek_relative() {
+}
+
+#[test]
+#[unstable(feature = "seek_convenience", issue = "59359")]
```
Same here.

LukasKalbertodt reviewed (changes requested):
Oops, I found another minor thing °_° 
Sorry for not noticing that in my last review.

LukasKalbertodt commented:
Looks good to me! 

@bors r+

Thanks for sticking with this PR for so long! :)

bors commented:
:pushpin: Commit 246d3271fe5fa180713b16aa33dacc6aeca46851 has been approved by `LukasKalbertodt`

<!-- @bors r=LukasKalbertodt 246d3271fe5fa180713b16aa33dacc6aeca46851 -->
<!-- homu: {"type":"Approved","sha":"246d3271fe5fa180713b16aa33dacc6aeca46851","approver":"LukasKalbertodt"} -->

bors commented:
:hourglass: Testing commit 246d3271fe5fa180713b16aa33dacc6aeca46851 with merge 9fe551ae49289ce6f693ca0dabf4c9c15164f67d...
<!-- homu: {"type":"BuildStarted","head_sha":"246d3271fe5fa180713b16aa33dacc6aeca46851","merge_sha":"9fe551ae49289ce6f693ca0dabf4c9c15164f67d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1081431102), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36467)
Approved by: LukasKalbertodt
Pushing 9fe551ae49289ce6f693ca0dabf4c9c15164f67d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"LukasKalbertodt","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36467","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1081431102"},"merge_sha":"9fe551ae49289ce6f693ca0dabf4c9c15164f67d"} -->
